---
title: "P/BIO381 Tutorials: Working with RNA-seq data"
subtitle: "RNAseq Days 4 and 5"
date: 'October 11 and 13, 2021'
output:
  prettydoc::html_pretty:
    theme: cayman
fontsize: 18pt
---

### Learning Objectives

1. Test for differences in gene expression in the _Acartia tonsa_ reciprocal transplant data set (from the Brennan et al paper we read!)!
2. Gain comfort and familarity working in R.
3. Gain comfort and familarity working with [DESeq2](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html) for analysis and visualization.


## 0. Installing and loading packages or libraries into your R working environment and importing the data into DESeq.
You should have the code below based on our live coding in class on 10/6 and 10/11. Hopefully your code is even more annotated!

``` R
# Class play time with RT tonsa data on 10/6/21
# by Melissa Pespeni


## Set your working directory
setwd("~/github/hudsonica")

getwd()
# [1] "/Users/mpespeni/github/hudsonica"

## Import the libraries that we're likely to need in this session
library(DESeq2)
library(dplyr)
library(tidyr)
library(ggplot2)
library(scales)
library(ggpubr)
library(wesanderson)
library(vsn)  ### First: BiocManager::install("vsn") AND BiocManager::install("hexbin")

# Import the counts matrix
countsTable <- read.table("DE_counts_F1.txt", header=TRUE, row.names=1)
head(countsTable)
dim(countsTable)
#[1] 24362    16
countsTableRound <- round(countsTable) # bc DESeq2 doesn't like decimals (and Salmon outputs data with decimals)
head(countsTableRound)

#import the sample discription table
conds <- read.delim("RT_tonsa_F1_samples.txt", header=TRUE, stringsAsFactors = TRUE, row.names=1)
head(conds)


# Continued 10/11/21

# Let's see how many reads we have from each sample
colSums(countsTableRound)
mean(colSums(countsTableRound))
barplot(colSums(countsTableRound), names.arg=colnames(countsTableRound),cex.names=0.5, las=3,ylim=c(0,20000000))
abline(h=mean(colSums(countsTableRound)), col="blue", lwd=2)

# the average number of counts per gene
rowSums(countsTableRound)
mean(rowSums(countsTableRound)) # [1] 11930.81
median(rowSums(countsTableRound)) # [1] 2226

apply(countsTableRound,2,mean) # 2 in the apply function does the action across columns
apply(countsTableRound,1,mean) # 1 in the apply function does the action across rows
hist(apply(countsTableRound,1,mean),xlim=c(50000,150000), ylim=c(0,10),breaks=1000)


#### Creat a DESeq object and definee the experimental design here with the tilda

dds <- DESeqDataSetFromMatrix(countData = countsTableRound, colData=conds, 
                              design= ~ line + environment + line:environment)

dim(dds)

# Filter out genes with too few reads - keep reads with average > 10 reads per sample
dds <- dds[rowSums(counts(dds)) >160]
dim(dds)

# Run the DESeq model to test for differential gene expression
dds <- DESeq(dds)

# List the results you've generated
resultsNames(dds)
# [1] "Intercept"                  "line_combined_vs_ambient"   "environment_HH_vs_AA"      
# [4] "linecombined.environmentHH"


###############################################################
# Let's start with a PCA to visualize global gene expression patterns
vsd <- vst(dds, blind=FALSE)

data <- plotPCA(vsd, intgroup=c("line","environment"), returnData=TRUE)
percentVar <- round(100 * attr(data,"percentVar"))

ggplot(data, aes(PC1,PC2, color=environment, shape=line)) +
  geom_point(size=4, alpha=0.85) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) +
  theme_minimal()
# What patterns do we see? Clustering by groups, line and environment. 
# What gene expression results do we expect for each factor, main effects and/or interactions?
###############################################################


###### Continue on dds model from before the PCA
###### Order and summarize the results from specific contrasts

resInteraction <- results(dds, alpha=0.05)
resInteraction <- resInteraction[order(resInteraction$padj),]
head(resInteraction)  
# log2 fold change (MLE): linecombined.environmentHH 
# Wald test p-value: linecombined.environmentHH 
# DataFrame with 6 rows and 6 columns
# baseMean   log2FoldChange             lfcSE
# <numeric>        <numeric>         <numeric>
#   TRINITY_DN115950_c0_g1 2245.96833710016 4.05236657027431  0.35849009363941
# TRINITY_DN131561_c0_g1 3375.98585631527 4.64570323588784 0.439847092095126
# TRINITY_DN137662_c0_g1 16743.2339417793 4.90199502543358 0.474582668629774
# TRINITY_DN149842_c8_g4 25971.8234076635 4.27274070902778 0.420809331610128
# TRINITY_DN129565_c0_g3 24258.7559645763 4.30552583036551 0.426036869169939
# TRINITY_DN129401_c0_g5 11712.3126703744 4.46354575123161 0.446093532600069
# stat               pvalue                 padj
# <numeric>            <numeric>            <numeric>
#   TRINITY_DN115950_c0_g1 11.3039848022981 1.25395982311055e-29 2.99445605758799e-25
# TRINITY_DN131561_c0_g1 10.5620869601728 4.46620142043715e-26 5.33264449600196e-22
# TRINITY_DN137662_c0_g1  10.329064564424  5.2065839941322e-25 4.14444085932923e-21
# TRINITY_DN149842_c8_g4 10.1536263292431 3.19274527294018e-24 1.90606892794529e-20
# TRINITY_DN129565_c0_g3  10.105993499469 5.19661481660706e-24 2.48190323641153e-20
# TRINITY_DN129401_c0_g5  10.005851744175 1.43650165401076e-23 5.71727658296281e-20

summary(resInteraction)
# out of 24362 with nonzero total read count
# adjusted p-value < 0.05
# LFC > 0 (up)       : 2839, 12%
# LFC < 0 (down)     : 1053, 4.3%
# outliers [1]       : 9, 0.037%
# low counts [2]     : 473, 1.9%

# About 16% of genes tested show a significant interaction!

```


## 2. Run DESeq to test for differences in gene expression

DESeq uses a negative binomial GLM, which is used for modeling count variables, usually for over-dispersed count outcome variables like gene expression data. For example, considering one gene, one sample may have 10,000 read counts (i.e., reads that mapped to that gene from that sample) while another sample may have 10 read counts that map to that gene.

DESeq has two ways to test for significance, using the Wald test (standard) or using the Likelihood Ratio Test (LRT; useful for study designs where there may be an interaction between two factors, in this case line and environment). Using the LRT, however, we have to test for each effect separately. See code below.

You can copy and paste the code below, but take the opportunity to annotate the code.
```R

#######################
############################################## TEST FOR EFFECT OF ENVIRONMENT
#######################

dds <- DESeqDataSetFromMatrix(countData = countsTableRound, colData = conds, 
                              design = ~ line + environment)

dds <- DESeq(dds, test="LRT", reduced=~line)
# List the results you've generated
resultsNames(dds)

# Order and list and summarize results from specific contrasts
resEnv <- results(dds, alpha = 0.05)
resEnv <- resEnv[order(resEnv$padj),]
head(resEnv)

summary(resEnv)

resEnv <- resEnv[!is.na(resEnv$padj),]

degsEnv <- row.names(resEnv[resEnv$padj < 0.05,]) 

```
# Now test for the other main effect, line:
``` R
#######################
##############################################  TEST FOR EFFECT OF LINE
#######################

dds <- DESeqDataSetFromMatrix(countData = countsTableRound, colData = conds, 
                              design = ~ environment + line)

dds <- DESeq(dds, test="LRT", reduced=~environment)
resultsNames(dds)

resLine <- results(dds, alpha = 0.05)
resLine <- resLine[order(resLine$padj),]
head(resLine)


summary(resLine)


resLine <- resLine[!is.na(resLine$padj),]

degsline <- row.names(resLine[resLine$padj < 0.05,])

```
# Now test for an interaction:
```R

#######################
##############################################  TEST FOR INTERACTION
#######################

dds <- DESeqDataSetFromMatrix(countData = countsTableRound, colData = conds, 
                              design = ~ environment + line + environment:line)

dds <- DESeq(dds, test="LRT", reduced=~environment + line)
resultsNames(dds)

resInt <- results(dds, alpha = 0.05)
resInt <- resInt[order(resInt$padj),]
head(resInt)

summary(resInt)


resInt <- resInt[!is.na(resInt$padj),]

degsInt <- row.names(resInt[resInt$padj < 0.05,])

```

## Data visualization - Are our statistics working?
```R
### Plot Individual genes ### 

# Counts of specific top interaction gene! (important validatition that the normalization, model is working)
d <-plotCounts(dds, gene="TRINITY_DN138549_c1_g2", intgroup = (c("line","environment")), returnData=TRUE)
d

p <-ggplot(d, aes(x=environment, y=count, color=line, shape=line, group=line)) + 
  theme_minimal() + theme(text = element_text(size=20), panel.grid.major=element_line(colour="grey"))
p <- p + geom_point(position=position_jitter(w=0.2,h=0), size=3)
p <- p + stat_summary(fun = mean, geom = "line")
p <- p + stat_summary(fun = mean, geom = "point", size=5, alpha=0.7) 
p

```

## More Data Visualization: Venn Diagrams, we finally get to make one ourselves!
We'll use the Eulerr package because we all know how nice it is to have the circle scaled. For more info, read [here](https://cran.r-project.org/web/packages/eulerr/vignettes/introduction.html).

```R

#######################
############################################## PLOT OVERLAPPING DEGS IN VENN DIAGRAM
#######################

library(eulerr)

# Total
length(degsEnv)  # 448
length(degsline)  # 226
length(degsInt)  # 3854

# Intersections
length(intersect(degsEnv,degsline))  # 37
length(intersect(degsEnv,degsInt))  # 44
length(intersect(degsInt,degsline))  # 34

intEL <- intersect(degsEnv,degsline)
length(intersect(degsInt,intEL)) # 7

# Number unique
448-44-37-7 # 360
226-37-34-7 # 148
3854-44-34-7 # 3769


fit1 <- euler(c("Env" = 360, "Line" = 148, "Interaction" = 3769, "Env&Line" = 37, "Env&Interaction" = 44, "Line&Interaction" = 34, "Env&Line&Interaction" = 7))

plot(fit1,  lty = 1:3, quantities = TRUE)

plot(fit1, quantities = TRUE, fill = "transparent",
     lty = 1:3,
     labels = list(font = 4))
     
```

## Make a heat map of the top differentially expressed genes
As with the PCA, we'll use the `vsd` function again. `vst` is a transformation implemented in DESeq2, which is "roughly similar to putting the data on the log2 scale, while also dealing with the sampling variability of low counts" (according the the package [manual](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html)). It uses the design formula to calculate the within-group variability (if blind=FALSE) or the across-all-samples variability (if blind=TRUE).

```R
# Heatmap of top 20 genes sorted by pvalue

library(pheatmap)

# By environment

topgenes <- head(rownames(resInt),20)
mat <- assay(vsd)[topgenes,]
mat <- mat - rowMeans(mat)
df <- as.data.frame(colData(dds)[,c("line","environment")])
pheatmap(mat, annotation_col=df)

# By line

topgenes <- head(rownames(resLine),20)
mat <- assay(vsd)[topgenes,]
mat <- mat - rowMeans(mat)
df <- as.data.frame(colData(dds)[,c("line","environment")])
pheatmap(mat, annotation_col=df)
```

     