---
title: "P/BIO381 Tutorials: Working with RNA-seq data - Day 3"
date: 'February 5, 2018'
output:
  prettydoc::html_pretty:
    theme: cayman
fontsize: 18pt
---

# Learning Objectives for 2/5/18

1. To continue to reinforce our understanding of the general work flow or "pipeline" for processing and analyzing RNAseq data AND review vim and screen tricks
2. To check that the mapping (a.k.a. aligning) of each set of cleaned reads to a reference transcriptome successfully completed (Program: bwa; Input: .fastq; Output: .sam).
3. To learn what a sequence alignment file is and how we can extract gene expression and genetic variant data from these files.
4. To extract read count data from individual .sam files (i.e. the number of reads that map (align) to each "gene") and see how they can be stitched together with bash scripting.
5. Get DESeq2 installed in R.

## Review ```vim``` and ```screen``` tricks

```vim``` is a text editor that can be used from the command line.  It can be quite powerful for editing once you know a few tricks.  We started with the basics: To open the file ```trim_myfile.sh```

```bash
vim trim_myfile.sh 
```

To edit or "insert" text, type ```i```

To exit the edit mode, hit ```esc```

To save your changes, you "write" the file with ```:w``` .  Note that you start all commands in vim with a colon. To quit the program, ```:q``` . 

```screen``` is another fantastically useful program that allows programs or scripts to run in the background if you "detach" from the screen.  To open a screen window, type ```screen```.  Now you have your command prompt and can start a script or enter a command.  To "detach", type ```ctrl``` plus ```a``` plus ```d``` . To reattach, type ```screen -r``` and the session ID you want to enter.  

## You’ve made a Sequence AlignMent (.sam) file!

### Let's take a look!  Try `head` and `tail`.

```bash
tail -n 100 YOURFILENAME.sam > tail.sam
vim tail.sam

:set nowrap
```

A SAM file is a tab delimited text file that stores information about the alignment of reads in a FASTQ file to a reference genome or transcriptome. For each read in a FASTQ file, there’s a line in the SAM file that includes

- the read, aka. query, name,
- a FLAG (number with information about mapping success and orientation and whether the read is the left or right read),
- the reference sequence name to which the read mapped
- the leftmost position in the reference where the read mapped
- the mapping quality (Phred-scaled)
- a CIGAR string that gives alignment information (how many bases Match (M), where there’s an Insertion (I) or Deletion (D))
- an ‘=’, mate position, inferred insert size (columns 7,8,9),
- the query sequence and Phred-scaled quality from the FASTQ file (columns 10 and 11),
- then Lots of good information in TAGS at the end, if the read mapped, including whether it is a unique read (XT:A:U), the number of best hits (X0:i:1), the number of suboptimal hits (X1:i:0).

The left (R1) and right (R2) reads alternate through the file. SAM files usually have a header section with general information where each line starts with the ‘@’ symbol. SAM and BAM files contain the same information; SAM is human readable and BAM is in binary code and therefore has a smaller file size.

Find the official Sequence AlignMent file documentation can be found [here](http://samtools.github.io/hts-specs/SAMv1.pdf).

[This BWA man page](http://bio-bwa.sourceforge.net/bwa.shtml) also discusses SAM alignment format and BWA specific optional fields.

- [Some FLAGs to know](http://seqanswers.com/forums/showthread.php?t=17314) - for example what do the numbers in the second column of data mean? [Here’s a SAM FLAG decoder](https://broadinstitute.github.io/picard/explain-flags.html) by the Broad Institute.
- What about the map quality score, MapQ? That’s important! [Here’s a reference](http://www.acgt.me/blog/2014/12/16/understanding-mapq-scores-in-sam-files-does-37-42).

#### How can we get a summary of how well our reads mapped to the reference?

```bash
samtools flagstat *.sam
```

#### Let’s see how many of our reads map uniquely.

Why is it important to consider whether a read maps uniquely (i.e., to one place in the transcriptome) for gene expression studies?

```bash
$ grep -c XT:A:U YOURFILENAME.sam 
1177827

$ grep -c X0:i:1 YOURFILENAME.sam
1182952
```

You can check a number of other elements, total number of reads, search for the various flags…

## Extract read counts from the .sam file from each sample

We will use a custom python script (by my friend Dan Barshis and published with the Simple Fool’s Guide to Population Genomics) called **countxpression.py**. This script will take any number of input *.sam files and, for each .sam file, extract the number of reads that map to each gene (i.e. the “counts”). It will also generate a summary output of useful information including proportion of quality read alignments. The script requires 4 input variables: mapqualitythreshold, lengththreshold, outputstatsfilename, anynumberofinputfiles.

```bash
cd /data/scripts
cp countxpression_pe.py ~/scripts      #or copy to your directory with the .sam file

python countxpression_pe.py 20 35 countstatssummary.txt YOURFILENAME.sam
```

This python script will generate two files: a .txt file you named (3rd argument you passed the script) and a counts .txt file that includes the number of uniquely mapped reads to each gene in our transcriptome.

Below is what the ```*_counts.txt```  should look like:

```bash
$ head NC_AD4_M3_bwaaln_counts.txt
ContigName  UniqueTotReads  MultiTotReads   totalreadsgoodmapped
OTAU000001-RA   11  207 218
OTAU000002-RA   982 49  1031
OTAU000003-RA   867 0   867
OTAU000004-RA   338 0   338
OTAU000005-RA   154 0   154
OTAU000006-RA   26  0   26
OTAU000007-RA   17  0   17
OTAU000008-RA   1017    55  1072
OTAU000009-RA   1984    0   1984
```

###Bash scripting to stitch together the counts files extracted from each .sam file

Once we have all the read counts extracted from each .sam file and **in one directory**, we can stitch them together with some bash scripting!  Then we'll have a data matrix that we can use to analyze global gene expression patterns! **You won't need to do this now**, but we can review the code if we have time in class or you can at another time.

```bash
# This loop takes the second column of data and renames the file to a shorter version of itself
for filename in *counts.txt; do
    myShort=`echo $filename | cut -c1-11` 
    echo "$myShort" > $myShort"_uniqmaps.txt"    
    cut -f 2 "$filename" > $myShort"_uniqmaps.txt"  
done 
# makes many individual files, but they don't have the header inserted

# This loop uses the tail command to get rid of the the first line
for filename in *_uniqmaps.txt; do
    tail -n +2 -- "$filename" > $filename"_uniqmapsNH.txt"  
done 

# This loop inserts the shortened version of the filename as the first line using the echo (print) and cat functions
for filename in *_uniqmapsNH.txt; do (myShort=`echo $filename | cut -c1-11`;echo "$myShort"; cat $filename) > tmp; mv tmp $filename; done

# This combines all the single column datafiles into one!
paste *_uniqmapsNH.txt > allcountsdata.txt

# Add row/gene names to table by cutting and pasting in the first column from one of your counts files.
cut -f 1 38_6-24_S_5_bwaaln_counts.txt | paste - allcountsdata.txt > allcountsdataRN.txt

# Change the name back to allcountsdata.txt
mv allcountsdataRN.txt allcountsdata.txt

# Check out your data file!
vim allcountsdata.txt

# to view with tabs aligned.
:set nowrap  

# clean up files, get rid of intermediate files
rm *uniqmaps*
```

### Let’s install DESeq2 in R studio and look at a script and example data file.

[Here’s](https://bioconductor.org/packages/release/bioc/html/DESeq2.html) the package website with installation instructions, manual, tutorials, etc.

Love MI, Huber W and Anders S (2014). “Moderated estimation of fold change and dispersion for RNA-seq data with DESeq2.” *Genome Biology*, **15**, pp. 550. [doi: 10.1186/s13059-014-0550-8](http://doi.org/10.1186/s13059-014-0550-8).

```R
source("https://bioconductor.org/biocLite.R")
biocLite("DESeq2")
```