---
title: "Ecological Genomics Tutorials: Population & Landscape Genomics 7"
date: 'October 02, 2023'
output:
  prettydoc::html_pretty:
    theme: cayman
fontsize: 18pt
---

# Learning Objectives for 10/02/23

1. Review the population structure results
2. Perform a scan for selection and identify contigs with outlier loci
3. Identify and visualize outlier loci 


### 1. What have we learned about the genetic structure from the PCA and admixture plots?

Let's take a look at the PCA and admixture figures posted to the Slack #coding channel.

- *What does the PCA seem to be telling us?*

- *What different picture does the admixture plot reveal?  How does it relate to the PCA?*

- *Would we want to look for higher levels of K in the admixture analysis?  How do we do that??*


### 2. How has selection acted to drive divergence of individual loci in excess of the population structure we've observed?

When selection acts in response to local environmental conditions, we observe an excess of population structure at certain loci. This can be thought of as Fst at a single locus exceeding some background level of divergence that exists across the genome as a whole. We call these "Fst outliers" and identifying these outliers is a major goal in ecological genomic studies. But, a major challenge is how best to control for background population structure -- in other words, how can we sift out the "outliers" from the rest of the population structure in the genome?

* One approach to identifying Fst outliers is using genetic PCA to (a) identify the major axes of population structure, and then (b) find loci that "load" very strongly on these axes, indicating their exceptional divergence is probably driven by the action of selection in excess of genetic drift.

* We can run a scan for Fst outliers using `pcANGSD`, just like we did for the genetic PCA, but here we initiate a selection scan for loci that are exceptionally divergent along one or more of the inferred genetic PC axes.  This method is especially helpful when it's hard to define what are genetic "populations", since it uses the genetic PCA to infer the genetic structure directly.  The method employed in `pcANGSD` follows the "FastPCA" selection scan method described by [Galinsky et al. 2015](https://www.cell.com/ajhg/fulltext/S0002-9297(16)00003-3) 

* The approach essentially uses the SNP weights or "loadings" on a given genetic PC axis to determine the strength of association, and if it exceeds the expectation due to neutral drift.

* Let's write a bash script called `pcANGSD_selection.sh` that includes the following code. Recall that `pcANGSD` uses genotype likelihoods in "beagle" format, which have been calculated for you [(see last tutorial)](https://pespenilab.github.io/Ecological-Genomics/Fall2023/tutorials/2023-09-27_PopGenomics_Day6.html) and can be found here:
`/netfiles/ecogen/PopulationGenomics/ANGSD/`

Here's a start to our script:

```
INPUT=""

OUTPUT=

SUFFIX="allRS_poly"

cp ${INPUT}/allRS_bam.list ${OUTPUT}

# Activates the pcANGSD environment and run the selection scan:

source /data/popgen/pcangsd/venv/bin/activate
        
pcangsd -b ${INPUT}/${SUFFIX}.beagle.gz \
	-o ${OUTPUT}/${SUFFIX} \
	-e <what number should we use here?>
	--selection \
	--minMaf 0.05 \
	--sites_save \
	--snp_weights \
	--threads 1
	
```
Note the `minMaf 0.05` option tests only those loci that have a minor allele frequency of 0.05 of greater. That's because there's little statistical power for testing association of alleles that are very rare (<5% in the sample).

It should run quickly! Once finished, you'll the see a new set of files, including:

| File name | contents |
|--------|-------------------|
| allRS_poly.selection.npy  | selection scores for each locus on each tested PC axis |
| allRS_poly.weights.npy | weights that show how strongly each locus "loads" on each PC axis |
| allRS_poly.sites | for each locus shows whether it got tested (1) or not (0)  |

You'll also want to get a file with minor allele frequencies ("maf") that ANGSD makes when first estimating genotype likelihoods. You have one of these files from when you worked on your focal pops, but we need one for the "allRS_poly" run of all red spruce individuals for just polymorphic sites.  I've provided this here:

`/netfiles/ecogen/PopulationGenomics/ANGSD/allRS_poly.mafs.gz`

We need to combine the "mafs" file and the "sites" file to create a file that has the contig and position info for each locus that got tested for selection.  Here's a few lines of bash code to create this file:

1. First, we need to add a header to the "sites" file:

`sed -i '1i kept_sites' ~/myresults/ANGSD/allRS_poly.sites`

2. Next, we need to get the contig and position info out of the "mafs" file and combine it with the "sites" file:

`zcat /netfiles/ecogen/PopulationGenomics/ANGSD/allRS_poly.mafs.gz | cut -f1-7 | paste - ~/myresults/ANGSD/allRS_poly.sites >allRS_poly_mafs.sites`

Now we have all the files we need!  Be sure to transfer the 3 listed above, plus the new `allRS_poly_mafs.sites` you just made, using **FileZilla** and save into your repo on your laptop.  When you have these files transferred (don't forget where you saved them to on your laptop!), open up **RStudio** and let's start making some figures!



### 3. Idenitfy and visualize outlier loci

We'll use R to import the selection scan results and associated meta-data, assign p-values for each tested locus, and visualize the results!  

Just a reminder, the following is R code, not bash. ;)

```

###################################
#  Selection scans for red spruce #
###################################

library(RcppCNPy) # for reading python numpy (.npy) files

setwd("~/Documents/Github/Ecological_Genomics/Fall_2023/pcangsd/")

list.files()

### read in selection statistics (these are chi^2 distributed)

s<-npyLoad("allRS_poly.selection.npy")

# convert test statistic to p-value
pval <- as.data.frame(1-pchisq(s,1))
names(pval) = "p_PC1"

## read positions
p <- read.table("allRS_poly_mafs.sites",sep="\t",header=T, stringsAsFactors=T)
dim(p)

p_filtered = p[which(p$kept_sites==1),]
dim(p_filtered)

# How many sites got filtered out when testing for selection? Why?

## make manhattan plot
plot(-log10(pval$p_PC1),
  col=p_filtered$chromo,
  xlab="Position",
  ylab="-log10(p-value)",
  main="Selection outliers: pcANGSD e=1 (K2)")

# We can zoom in if there's something interesting near a position...

plot(-log10(pval$p_PC1[2e05:2.01e05]),
  col=p_filtered$chromo, 
  xlab="Position", 
  ylab="-log10(p-value)", 
  main="Selection outliers: pcANGSD e=1 (K2)")

# get the contig with the lowest p-value for selection
sel_contig <- p_filtered[which(pval==min(pval$p_PC1)),c("chromo","position")]
sel_contig

# get all the outliers with p-values below some cutoff
cutoff=1e-3   # equals a 1 in 5,000 probability
outlier_contigs <- p_filtered[which(pval<cutoff),c("chromo","position")]
outlier_contigs

# how many outlier loci < the cutoff?
dim(outlier_contigs)[1]

# how many unique contigs harbor outlier loci?
length(unique(outlier_contigs$chromo))

```

How do we find out what functional genes are contained on the unique contigs harboring outliers?

We can use the *Picea abies* reference genome annotation to get the genes based on the outlier contigs, then test for enrichment of gene function using available public databases.

First, export your unique outlier contigs in R:

```
write.table(unique(outlier_contigs$chromo),
  "allRS_poly_PC1_outlier_contigs.txt", 
  sep="\t",
  quote=F,
  row.names=F,
  col.names=F)
```

Transfer back to the server and then grep out the gene IDs. We can do this with a series of piped commands like so:

- Define your paths to the ref genome annotation on the server and your newly made outlier loci file

- Use `zcat` to open the annotation without unzipping it, then pipe to `grep` and use the `-f` flag to take the contig names from the outliers file -- this is a handy trick, and saves us a lot of time from having to manually input each contig one at a time and search!

- Pipe to a new `grep` to get just the contigs containing genes

- Pipe the list of gene IDs to take just the unique ones, cut out the 9th column (containing the gene IDs), and get rid of the annoying "ID= " portion of each gene ID

Just a reminder, the following is bach code, not R ;)

```
ANNOT="/netfiles/ecogen/PopulationGenomics/ref_genome/annotation/Pabies1.0-all-cds.gff3.gz"

OUTLIERS=~/myresults/ANGSD/allRS_poly_PC1_outlier_contigs.txt

zcat ${ANNOT} | grep -f ${OUTLIERS} | grep "gene" | uniq | cut -f9 | sed "s/ID=//g"
```

Voila! You should now have a list of gene IDs printed to your screen. You can either copy these by highlighting with your mouse and clicking <ctrl><c> or you can save to an external file for later (using the `>outfile.txt` command)

Take your gene IDs and go to the [plantgenie](https://plantgenie.org) website. Here, you can create a gene list using your copied IDs, determine which genes they correspond to (not all wil be annotated!), and test for functional enrichment:

- In the upper left corner, see a circle with 3 red bars --> click it, and create a new gene list

- Paste your gene IDs into the list, and plantgenie will search the P. abies annotation and return all the known genes

- Click the "+" button to save these to the current gene list, then close out

- Go to the "Analysis Tools" tab on the top bar of the page, and choose "Enrichment"

- Plantgenie will now test for whether the annotation Gene Ontology (GO) and Protein Family (Pfam) functional descriptions assigned to each gene ID are over-represented compared to all other genes containing those GO and Pfam categories in the P. abies genome.

- Relate back to the biology!
  - *What sort of functional genes did you find are represented in your outliers?*
  - *Are any enriched for certain GO or Pfam functions? (focus on p-values << 0.01, or q-values < 0.05)*
  - *Remember the population structure axis (genetic PC) these loci are outliers along...are any of the enriched functions interesting in light of this structure? Any surprising?*
