---
title: "Population Genomics Tutorial 4: Estimating genomic diversity"
author: "SRK"
date: 'Fall 2025'
output:
  prettydoc::html_pretty:
    theme: cayman
fontsize: 18pt
---


## 1. Inference of population genomics from the aligned sequence data: should we call genotypes?

Many of the papers you'll read that do popgen on NGS data have a SNP calling step that results in a specific genotype being called for each SNP site for each individual. For example,

| SNP | Ind1 | Ind 2 |
|-----|------|-------|
| 1   | CC   | CT    |
| 2   | AG   | AA    |
| 3   | GT   | TT    |

But how do we know that Ind1 is homozygous at SNP-1 (CC) -- couldn't it be CT and we just didn't have enough coverage to observe the second allele?

The basic problem is that read data are counts that produce a binomial distribution of allele calls at a given site, and if you have few reads, you might by chance not observe the true genotype. So, what's the right thing to do?

As with almost anything in statistics, the right thing to do is not throw away that uncertainty, but instead incorporate it into your analysis. That's what we're going to do...

### Genotype-free population genetics using genotype likelihoods

A growing movement in popgen analysis of NGS data is embracing the use of genotype likelihoods to calculate stats based on each individual having a likelihood (probability) of being each genotype.

**A genotype likelihood (GL) is essentially the probability of observing the sequencing data (reads containing a particular base), given the genotype of the individual at that site.**

These probabilities are modeled explicitly in the calculation of population diversty stats like Theta-pi, Tajima's D, Fst, PCA, etc...; thus not throwing out any precious data, but also making fewer assumptions about the true (unknown) genotype at each locus

-   We're going to use this approach with the program 'ANGSD', which stands for 'Analysis of Next Generation Sequence Data'

-   This approach was pioneered by Rasmus Nielsen, published originally in [Korneliussen et al. 2014](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-014-0356-4).

-   [ANGSD has a user's manual (it's a work in progress...)](http://www.popgen.dk/angsd/index.php/ANGSD)

----------
#### The basic work flow of ANGSD goes like this:

1.  Create a list of bam files for the samples you want to analyze
2.  Estimate genotype likelihoods (GL's) and allele frequencies after filtering to minimize noise
3.  Use GL's to:
    (a) estimate the site frequency spectrum (SFS)

    (b) estimate nucleotide diversities and neutrality stats (Thetas, Tajima's D, ...)


#### 1. In your `myscripts/` folder, create a new text file called `ANGSD.sh`

Create the header for your inputs and outputs

```         
cd ~/projects/eco_genomics_2025/population_genomics/myresults

mkdir ANGSD

INPUT="/gpfs1/cl/ecogen/pbio6800/PopulationGenomics/bams"

OUTPUT="ANGSD"

REF="/gpfs1/cl/ecogen/pbio6800/PopulationGenomics/ref_genome/Pmariana1.0-genome_reduced.fa"

MYPOP=""

ls ${INPUT}/${MYPOP}*sorted.rmdup.bam >${OUTPUT}/${MYPOP}_bam.list

```
Save your file and try running it at the command line.

*Check your output bamlist to see it was made properly!* 

-  Where would you look for this file? (Hint, refer back to the `ls` command that makes it). 
-  How would you verify its contents? (hint: use `head` or `cat` or even open in a text file in RStudio OOD)


#### 2. Open your `ANGSD.sh` script back up in RStudio

Estimate your GL's and allele freqs, optionally filtering for base and mapping quality, sequencing depth, SNP probability, minor allele frequency, etc.

Add the following code chunk at the bottom of your script:

```         
# File suffix to distinguish analysis choices, like "All" or "Poly" depending on whether you're analyzing all the sites or just the polymorphic ones
SUFFIX=""

# Estimating GL's and allele frequencies for all sites with ANGSD

######################

ANGSD -b ${OUTPUT}/${MYPOP}_bam.list \
-ref ${REF} -anc ${REF} \
-out ${OUTPUT}/${MYPOP}_${SUFFIX} \
-nThreads 10 \
-remove_bads 1 \
-C 50 \
-baq 1 \
-minMapQ 20 \
-minQ 20 \
-GL 2 \
-doSaf 1 \
##### below filters require `do-Counts`
#-doCounts 1 \
#-minInd 4 \
#-setMinDepthInd 1 \
#-setMaxDepthInd 40 \
#-setMinDepth 20 \
#-skipTriallelic 1 \
#-doMajorMinor 1 \
##### below filters require `doMaf`
#-doMaf 1 \
#-SNP_pval 1e-6 \
#-minMaf 0.01
```

What do all these options mean?

| Option             | Description                                                         |
|--------------------|---------------------------------------------------------------------|
| -nThreads 10       | how many cpus to use -- in this example, 10 cpus!                   |
| -remove_bads 1     | remove reads flagged as 'bad' by samtools                           |
| -C 50              | enforce downgrading of map quality if contains excessive mismatches |
| -baq 1             | estimates base alignment qualities for bases around indels          |
| -minMapQ 20        | threshold for minimum read mapping quality (Phred)                  |
| -minQ 20           | threshold for minimum base quality (Phred)                          |
| -GL 2              | calculate genotype likelihoods (GL) using the GATK formula          |
| -doSaf 1           | output allele frequency likelihoods for each site                   |
| -doCounts 1        | output allele counts for each site                                  |
| -minInd 4          | min number of individuals to keep a site (see also ext 2 filters)   |
| -setMinDepthInd 1  | min read depth for an individual to count towards a site            |
| -setMaxDepthInd 40 | max read depth for an individual to count towards a site            |
| -setMinDepth 20    | min read depth across ALL individual to keep a site                 |
| -skipTriallelic 1  | don't use sites with \>2 alleles                                    |
| -doMajorMinor 1    | fix major and minor alleles the same across all samples             |
| -doMaf 4           | Use the reference allele as the major allele.                       |
| -SNP_pval 1e-6     | Keep only site highly likely to be polymorphic (SNPs)               |
| -minMaf 0.01       | Keep only sites with minor allele freq > some proportion (e.g., 1%) |

**NOTES**

-   If you want to restrict the estimation of the genotype likelihoods to a particular set of sites you're interested in, add the option `-sites selected_sites.txt` (tab delimited file with the position of the site in column 1 and the chromosome in column 2) or use `-rf selected_chromosome.chrs` (if listing just the unique "chromosomes" or contigs you want to anlayze)
-   Some popgen stats you want to estimate only the polymorphic sites; for this you should include the `-SNP_pval 1e-6` option to eliminate monomorphic sites when calculating your GL's
-   There are good reasons to do it BOTH ways, with and without the `-SNP_pval 1e-6` option. Keeping the monomorphic sites is essential for getting proper estimates of nucleotide diversity and Tajima's D. But other analyses such as PCA or GWAS want only the SNPs.

*Save your script. Then run at the command line**

At the end of this run, you should have the following output files in your `myresults/ANGSD` directory:

```
9999_.saf.gz
9999_.saf.idx
9999_.saf.pos.gz
```

These "saf" files contain "site allele frequency" likelihoods, and are the info needed to estimate stats that depend on population allele frequencies, like nucleotide diversities, neutrality stats like Tajima's D, and population divergence stats like Fst. Each of these stats depends on the *SFS -- the Site Frequency Spectrum*

So, our workflow will be to (1) use the .saf files to estimate the SFS, and (2) to then use the SFS to estimate nucleotide diversity and associated statistics.


#### 3a. In your `myscripts` folder, create `ANGSD_doTheta.sh` to estimate the SFS and nucleotide diversity stats for your pop

Based on the saf.idx files from ANGSD GL calls, you can estimate the Site Frequency Spectrum (SFS), which is the precursor to many other analyses such as nucleotide diversities (as well as Fst, demographic history analysis, etc.)


```         
REPO="~/projects/eco_genomics_2025/population_genomics/"

REF="/gpfs1/cl/ecogen/pbio6800/PopulationGenomics/ref_genome/Pmariana1.0-genome_reduced.fa"

OUT="${REPO}/myresults/ANGSD"

MYPOP=""

SUFFIX=""   # All sites or just polymorphic sites?

#Estimation of the SFS for all sites using the FOLDED SFS
realSFS ${OUT}/${MYPOP}_${SUFFIX}.saf.idx \
        -maxIter 1000 \
        -tole 1e-6 \
        -P 1 \
        > ${OUT}/${MYPOP}_${SUFFIX}.sfs

```

#### 3b. After the SFS, estimate the theta diversity stats:

Once you have the SFS, you can estimate the thetas and neutrality stats by adding the following code chunk at the end of your `ANGSD_doTheta.sh` script:

```         
# Estimate thetas and stats using the SFS

realSFS saf2theta ${OUT}/${MYPOP}_${SUFFIX}.saf.idx \
        -sfs ${OUT}/${MYPOP}_${SUFFIX}.sfs \
        -outname ${OUT}/${MYPOP}_${SUFFIX}

thetaStat do_stat ${OUT}/${MYPOP}_${SUFFIX}.thetas.idx

```
If we wanted to analyze this on sliding windows, we could instead replace the above code chunk with the following:

```
# For sliding window analysis:

thetaStat do_stat ${OUT}/${MYPOP}_${SUFFIX}.thetas.idx \
-win 50000 \
-step 10000 \
-outnames ${OUT}/${MYPOP}_${SUFFIX}.thetasWindow.gz
```



#### Now we have some diversity results!

For the results files above, the first column of the results file (*.thetas.idx.pestPG) is formatted a bit funny and we don't really need it. There are also a bunch of other neutrality stats that we don't need right now. We can use the `cut` command to get rid of these extra columns. The below `cut` command retains columns 2-5, 9 and 14 


```         
cut -f2-5,9,14 ${OUT}/${MYPOP}_${SUFFIX}.thetas.idx.pestPG > ${OUT}/${MYPOP}_${SUFFIX}.thetas
```


The columns correspond to the following stats:

| Col | Statistic    | Description                                                                         |
|-----|--------------|-------------------------------------------------------------------------------------|
| 2   | Chr          | The chromosome or (more appropriately) contig being analyzed                        |
| 3   | WinCenter    | The center of the window or contig, in bp                                           |
| 4   | tW           | Watterson's Theta -- an estimate of nucleotide diversity based on segregating sites |
| 5   | tP           | Theta Pi -- estimate of nucleotide diversity based on pairwise divergence           |
| 9   | Tajima       | Tajima's D -- a neutrality stat that tests for the difference in tW-tP              |
| 14  | nSites       | The number of base pairs being analyzed along this stretch of the genome            |


### 2. Summarize diversity stats in R

We're now ready to use `RStudio OOD` to visualize the results:

Here's some basic R code to help you along:

```
setwd("") # set your path to your results folder in your repo where you saved your diversity stats file

list.files() # list out the files in this folder to make sure you're in the right spot.

# First let's read in the diversity stats
theta <- read.table("_.thetas",sep="\t",header=T)

theta$tWsite = theta$tW/theta$nSites #scales the theta-W by the number of sites
theta$tPsite = theta$tP/theta$nSites #scales the theta-Pi by the number of sites

summary(theta)

# You can order the contig list to show you the windows/contigs with the highest values of Tajima's D, or the lowest

head(theta[order(theta$Tajima, decreasing = TRUE),]) # top 10 Tajima's D values

head(theta[order(theta$Tajima, decreasing = FALSE),]) # bottom 10 Tajima's D values

#You can also look for windows/contigs that have combinations of high Tajima's D and low diversity -- these could represent outliers for selection!
#theta[which(theta$Tajima<1.5 & theta$tPsite<0.001),]

# We may want to visualize the site frequency spectrum (SFS) as a barplot of the site frequencies
# Be sure to replace "9999" with your pop code in the "main" legend below
barplot(sfs,xlab="Chromosomes",
        names=1:length(sfs),
        ylab="Proportions",
        main="Pop 9999 Site Frequency Spectrum",
        col='blue')
        

# We may also want to know how many sites are monomorphic (not variable):
sfs<-scan('9999_.sfs')
sfs<-sfs[-c(1,which(sfs==0))]
sfs<-sfs/sum(sfs)

# Put the nucleotide diversities, Tajima's D, and SFS into a 4-panel figure
par(mfrow=c(2,2))
hist(theta$tWsite, xlab="theta-W", main="Watterson's theta")
hist(theta$tPsite, xlab="theta-Pi", main="Pairwise Nucleotide Diversity")
hist(theta$Tajima, xlab="D", main="Tajima's D")
barplot(sfs,names=1:length(sfs),main='Site Frequency Spectrum')


# To reset the panel plotting, execute the line below:
dev.off()

# We could also plot the nucleotide diversities and Tajima's D as sliding windows along the genome:

library(ggplot2)

ggplot()

```




